monai/transforms/io/dictionary.py

```diff
@@ -127,22 +132,44 @@ def __call__(self, data, reader: Optional[ImageReader] = None):
 
         """
         d = dict(data)
+        image_shapes = []
+        # From python 3.6 and up, dictionaries are ordered, so we can assume that the dict orders we get
+        # are all in the same order, unless the user goes out of their way to shuffle them.
+        first_path = list(d.values())[0]
+
         for key, meta_key, meta_key_postfix in self.key_iterator(d, self.meta_keys, self.meta_key_postfix):
-            data = self._loader(d[key], reader)
+            image_data, _ = self._loader(d[key], reader)
+            # _loader returns a tuple of (ndarray, dict), we want the ndarray.shape
+            image_shapes.append(image_data[0].shape)
+
             if self._loader.image_only:
-                if not isinstance(data, np.ndarray):
+                if not isinstance(image_data, np.ndarray):
                     raise ValueError("loader must return a numpy array (because image_only=True was used).")
-                d[key] = data
+                d[key] = image_data
             else:
-                if not isinstance(data, (tuple, list)):
+                if not isinstance(image_data, (tuple, list)):
                     raise ValueError("loader must return a tuple or list (because image_only=False was used).")
-                d[key] = data[0]
-                if not isinstance(data[1], dict):
+                d[key] = image_data[0]
+                if not isinstance(image_data[1], dict):
                     raise ValueError("metadata must be a dict.")
                 meta_key = meta_key or f"{key}_{meta_key_postfix}"
                 if meta_key in d and not self.overwriting:
                     raise KeyError(f"Meta data with key {meta_key} already exists and overwriting=False.")
-                d[meta_key] = data[1]
+                d[meta_key] = image_data[1]
+
+        # Only warn the first time per sample to prevent spamming the logs
+        if self.log_warning and (first_path not in self.has_warned_about):
+            # all() is a lazy function, i.e. it will stop at the first instance of shapes mismatching
+            if not all(x == image_shapes[0] for x in image_shapes):
+                print("WARNING: Loaded image shapes do not match for input dictionary:")
+                print(dict(data))
+                print("\nResulting in image shapes:")
+                [print(f"\t{x}: {y.shape}") for x, y in zip(d.keys(), d.values())]
```

Comment:
I don't recommend using `for` expressions as shorthand for `for` loops, or any other expression in place of a statement. This adheres to Python style guides like [this](https://docs.python-guide.org/writing/style/) which I think is generally for us to follow.

Reply:
Would replacing this with
`'\n'.join([f"{key}: {value.shape}" for key, value in zip(d.keys(), d.values())])`
be more acceptable?

Or would you rather use a full `for` loop and build the string before printing?

Reply:
What you can do is build up a list of strings, say `msg`,  so here you would use a for loop at appends strings to `msg`, then the final line would be `msg = "\n".join(msg)`. You could then log or warn using `msg` as the full string.

Reply:
In moving to the `warnings` module, I have cleaned up the string building.
```python
warnings.warn(
    message="Loaded image shapes do not match for input dictionary:" +
                    f"\n{dict(data)}" +
                    "\nResulting in image shapes:" +
                    '\n'.join(f"\t{key}: {value.shape}" for key, value in d.items()) +
                    "If you do not wish to see this warning, pass `warn_on_shape_mismatch = False` to " +
                    "LoadImaged()",
    category=UserWarning
)
```

Is that more in line with your expectations?

Reply:
It's closer to what I would have done, we're now getting into personal style but it's much better. I would have done this which is maybe too busy looking:

```python
msg = [
  "Loaded image shapes do not match for input dictionary:",
  f"{dict(data)}",
  "Resulting in image shapes:" 
]
msg += [f"\t{key}: {value.shape}" for key, value in d.items() if hasattr(value, "shape")]
msg.append("If you do not wish to see this warning, pass `warn_on_shape_mismatch = False` to LoadImaged()")

warnings.warn(message="\n".join(msg), category=UserWarning)
```